### PR TITLE
Check for HTTP status 200 in fetchAndParse.

### PR DESCRIPTION
**Pull request by pphaneuf**
Also use a streaming JSON decoder for `fetchAndParse`, and ensure that the body is read to completion, so the connection can be reused.

### CODE REVIEW

File: go/client/logclient.go [hunk 1]
```diff
@@ -125,20 +125,21 @@ func (c *LogClient) fetchAndParse(uri string, res interface{}) error {
...
 	if err != nil {
 		return err
 	}
-	if err = json.Unmarshal(body, &res); err != nil {
+	defer resp.Body.Close()
+	// Make sure everything is read, so http.Client can reuse the connection.
+	defer ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode != 200 {
+		return errors.New("status code not 200")
```
Comment (phad) [inline go/client/logclient.go]: a better message might say what the status code actually was

File: go/client/logclient.go [hunk 2]
```diff
@@ -125,20 +125,21 @@ func (c *LogClient) fetchAndParse(uri string, res interface{}) error {
...
-			return err
-		}
-	}
 	if err != nil {
 		return err
 	}
-	if err = json.Unmarshal(body, &res); err != nil {
+	defer resp.Body.Close()
+	// Make sure everything is read, so http.Client can reuse the connection.
+	defer ioutil.ReadAll(resp.Body)
```
Comment (phad) [inline go/client/logclient.go]: is the point here that if json.NewDecoder...Decode doesn't run (or terminates prematurely) that the deferred 'ReadAll' gets to the end of the input regardless?

is that a noop in the case of JSON parsing out successfully?

File: go/client/logclient.go (hunk 1 above)
Comment (pphaneuf) [inline go/client/logclient.go]: Good point, done.

File: go/client/logclient.go (hunk 2 above)
Comment (pphaneuf) [inline go/client/logclient.go]: That's correct. In the case of the HTTP status not being 200, the JSON decoding won't even be attempted, so if we want to reuse the connection, we need to slurp whatever body came in. And when it is attempted, there might be trailing cruft at the end, and if the JSON decoding got everything, it just won't get anything.

File: go/client/logclient.go [hunk 3]
```diff
@@ -133,7 +133,7 @@ func (c *LogClient) fetchAndParse(uri string, res interface{}) error {
 	defer ioutil.ReadAll(resp.Body)
 
 	if resp.StatusCode != 200 {
-		return errors.New("status code not 200")
+		return errors.New("status code not 200: " + strconv.Itoa(resp.StatusCode))
```
Comment (phad) [inline go/client/logclient.go]: might fmt.Sprintf() be a bit shorter?

Comment (pphaneuf) [inline go/client/logclient.go]: `fmt.Sprintf()` was exactly the same, but there's a `fmt.Errorf()` that seems to fit perfectly. I made the message similar to another one at line 237, while I was there, for consistency.

File: go/client/logclient.go (hunk 2 above)
Comment (phad) [inline go/client/logclient.go]: Ah yes, that's ideal.

File: go/client/logclient.go (hunk 3 above)
Comment (phad) [inline go/client/logclient.go]: Yeah, that's better.

**Pull request merged**